ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java

```diff
@@ -435,15 +439,39 @@ private void transitionLayoutOnlyViewToNativeView(
     mTagsWithLayoutVisited.clear();
   }
 
-  private static boolean isLayoutOnlyAndCollapsable(@Nullable ReactStylesDiffMap props) {
+  private static boolean isCollapsableChildren(@Nullable ReactStylesDiffMap props) {
     if (props == null) {
       return true;
     }
 
+    if (props.hasKey(ViewProps.COLLAPSABLE_CHILDREN) && !props.getBoolean(ViewProps.COLLAPSABLE_CHILDREN, true)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private static boolean isCollapsable(@Nullable ReactStylesDiffMap props) {
+    if (props == null) {
+      return false;
+    }
+
     if (props.hasKey(ViewProps.COLLAPSABLE) && !props.getBoolean(ViewProps.COLLAPSABLE, true)) {
       return false;
     }
 
+    return true;
+  }
+
+  private static boolean isLayoutOnlyAndCollapsable(@Nullable ReactStylesDiffMap props) {
+    if (props == null) {
+      return true;
+    }
+
+    if (!isCollapsable(props) || !isCollapsableChildren(props)) {
```

Comment:
Haven't tested a case where that would happen but I'm pretty sure it would cause issues if we optimize out a view that we want to make sure it's child exists and map to the same index as Views in react js.